### DISCUSSION TITLE
What's the difference between cue export and cue eval?

### DISCUSSION THREAD

**Discussion by vikstrous2** [cue-lang/cue#1421] [Q&A] [answered]
They seem to do the same thing..?

**Comment by xinau** [comment, marked as answer]
Afaik. The difference between the two is that `export` requires the expression to be [concrete](https://pkg.go.dev/cuelang.org/go@v0.4.0/cue#Concrete) and [final](https://pkg.go.dev/cuelang.org/go@v0.4.0/cue#Final). This isn't the case for `eval`

This can be demonstrated with the following snippet.

```cue
output: string
```

resulting in the following output for `eval`

```console
$ cue eval test.cue
output: string
```

and an this error for `export`

```console
$ cue export test.cue
output: incomplete value string
```

Furthermore, `export` is intended to export the expression in a specific format like YAML or JSON.

**Reply by verdverm** [reply]
I'd also add that eval exists in a weird place between def and export. Eval is planned for depreciation iirc.

Also note, you can export incomplete values if the format is CUE

`cue export --out cue ...`

**Reply by mpvl** [reply]
Yeah, Tony is right. `cue eval` is a bit odd. We plan to deprecate `cue eval` in favor of a command-less `cue`, but we need to gather more use cases and feedback on what such an output should look like.

Also note that `cue eval` really is informational and may not even output correct cue (e.g. unresolvable references).

**Reply by wonderflow** [reply]
`cue eval` as a CLI command may be odd, but the Implementation function behind is really useful for tools to integrate. We combine CUE code fragment step by step, the result is something like a template, it could be validate and left some parameters as input for users. After user gives the input, we can export the final data efficiently. 😄
